test_runner/src/main/kotlin/ftl/shard/Shard.kt

```diff
@@ -1,123 +1,60 @@
         }
 
-        System.out.println("Slowest shard = " + timelist.max())
-        System.out.println("Total Time = $totalTime")
+        val testCount = testcases.size
 
-        // Ensure we don't create more VMs than requested. VM count per run should be <= testShards
-        if (testShardChunks.size > maxTestShards) {
-            fatalError("Calculated chunks $testShardChunks is > requested $maxTestShards testShards.")
-        }
-        if (testShardChunks.isEmpty()) fatalError("Failed to populate test shard chunks")
-    }
+        // If maxShards is infinite or we have more shards than tests, let's match it
+        val shardsCount = if (maxShards == -1 || maxShards > testCount) testCount else maxShards
 
-    /**
-     * Build shard by removing remaining methods that total to targetShardDuration
-     * At least one method per shard will always be returned, regardless of targetShardDuration.
-     *
-     * remainingMethods must be sorted in order of fastest execution time to slowest.
-     * remainingMethods.sortBy { it.time }
-     *
-     * Based on createConfigurableShard from Flank Java
-     * https://github.com/TestArmada/flank/blob/ceda6d2c3d9eb2a366f19b826e04289cd24bddf3/Flank/src/main/java/com/walmart/otto/shards/ShardCreator.java#L99
-     */
-    fun build(remainingMethods: MutableList<TestMethod>, targetShardDuration: Double): TestShard {
-        var timeBudget = targetShardDuration
-        var shardTime = 0.0
-
-        val testMethods = remainingMethods.iterator()
-        val shardTests = mutableListOf<TestMethod>()
-
-        while (testMethods.hasNext()) {
-            val test = testMethods.next()
-            val testWithinBudget = timeBudget - test.time >= 0
-
-            if (testWithinBudget) {
-                timeBudget -= test.time
-
-                shardTime += test.time
-                shardTests.add(test)
-                testMethods.remove()
-
-                continue
-            }
+        // Create the list of shards we will return
+        var shards = List(shardsCount) { TestShard(0.0, mutableListOf()) }
 
-            val noTestsAdded = timeBudget == targetShardDuration
-            val testOverBudget = test.time >= timeBudget
+        // We want to iterate over testcase going from higher to lower
+        testcases.sortByDescending { it.time }
 
-            if (noTestsAdded && testOverBudget) {
-                // Always add at least 1 test per shard regardless of budget
-                shardTime += test.time
-                shardTests.add(test)
-                testMethods.remove()
+        testcases.forEach {
 
-                return TestShard(shardTime, shardTests)
-            }
+            val tm = TestMethod(it.name, it.time)
+            val shard = shards[0]
+
+            shard.testMethods.add(tm)
+            shard.time += tm.time
+
+            // Sort the shards to keep the most empty shard first
+            shards = shards.sortedBy { it.time }
```

Comment:
it is shadowed here

![image](https://user-images.githubusercontent.com/1173057/48631619-f03ca680-e98c-11e8-9618-67a7383a9bdd.png)